### PR TITLE
Added insert_start function for Doubly Linked List

### LINKED ISSUE #77: Insert an item in front of a doubly linked list
Currently there is a `doubly_linked_list.py` file in dsalgo (not merged yet) folder but there is no way to insert an item in front of a given doubly linked list for example, if function name is `insert_start()` (this is preferred name) 

then:
- [ ] code for which items.insert_start([_this contains required object_]) will show up the new object added in front of the doubly linked list created previously in continuation with issue #67 
- [ ] make a pull request for queue branch not for master branch.
- [ ] update the queue branch with the master.

P.S : Refer [this](https://github.com/codesankalp/dsalgo/pull/57/commits/a9e3e248d7d5745b00661f46ed5324c9f0ea17e2) commit for doubly_linked_list.py file. 
or the issue #56 in continuation with issue #67

### PR DESCRIPTION
**Pull request by dennis-pg**
### Description

Addressing Issue #77 
Added required insert_start function

### Related Issue

** Please don't create a Pull request without creating an issue first **

Fixes #77 

### Type of Change:

- Code

### How Has This Been Tested?

Sanity checks ran

### Checklist:

- [x] I read the **CONTRIBUTING** document before making this pull request
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] My change requires change to documentation
- [ ] I have updated documentation accordingly
- [ ] I have updated the requirements file
- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: dsalgo/doubly_linked_list.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        current = self.head
+        while current:
+            item_val = current.data
+            current = current.next
+            yield item_val
+
+# Use items = doubly_linked_list()
+# then items.append_item() append elements here any number of times
+# and use print_forward() to print the output as below :
+# items.print_foward()
```
Comment (devkapilbansal) [inline dsalgo/doubly_linked_list.py:54]: Why these comments are here??

Comment (devkapilbansal) [inline dsalgo/doubly_linked_list.py:54]: Add a blank line after each file

File: dsalgo/doubly_linked_list.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        new_item.next = self.head
+        self.head = new_item
+        if self.tail is None:
+            self.tail = self.head
+        
+        self.count += 1
+
+
+    
+    def print_foward(self):
```
Comment (devkapilbansal) [inline dsalgo/doubly_linked_list.py:39]: If I am not wrong, only two blank lines are needed after each function

File: dsalgo/doubly_linked_list.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            self.tail = self.head
+        else:
+            new_item.prev = self.tail
+            self.tail.next = new_item
+            self.tail = new_item
+
+        self.count += 1
+
+    def insert_start(self, data):
+        #Insert item at the beginning
```
Comment (devkapilbansal) [inline dsalgo/doubly_linked_list.py:28]: I would have write it as `Insert a node in the beginning`

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
Thanks @dennis-pg see my request below.

**Comment by devkapilbansal** [comment]
@dennis-pg add test cases for this too

**Comment by devkapilbansal** [comment]
Also, @dennis-pg there are some more issues open related to it which are not assigned. Solve those too.

File: dsalgo/doubly_linked_list.py (hunk 1 above)
Comment (keshavgbpecdelhi) [reply dsalgo/doubly_linked_list.py:54]: To explain how to call these functions. Ok! I will remove it.

Comment (keshavgbpecdelhi) [reply dsalgo/doubly_linked_list.py:54]: Blank line, each file 🧐so I have to arrange comments with functions starting !?

File: dsalgo/doubly_linked_list.py (hunk 2 above)
Comment (keshavgbpecdelhi) [reply dsalgo/doubly_linked_list.py:39]: Yes, it's by mistake.

File: dsalgo/doubly_linked_list.py (hunk 3 above)
Comment (keshavgbpecdelhi) [reply dsalgo/doubly_linked_list.py]: Done at https://github.com/codesankalp/dsalgo/pull/78/files#diff-c1dabacd7e9736a181ab83bf83dabbbb

File: dsalgo/doubly_linked_list.py (hunk 1 above)
Comment (keshavgbpecdelhi) [reply dsalgo/doubly_linked_list.py]: Tried to resolve

Comment (keshavgbpecdelhi) [reply dsalgo/doubly_linked_list.py]: Can you check is it now correct @ https://github.com/codesankalp/dsalgo/pull/78/files#diff-c1dabacd7e9736a181ab83bf83dabbbb

**Comment by dennis-pg** [comment]
Okay, will have a look

**Comment by keshavgbpecdelhi** [comment]
@devkapilbansal  I did some changes in the code [here](https://github.com/codesankalp/dsalgo/pull/78/commits/00c099df16a6f43855617698657a3992d6514195)

File: dsalgo/doubly_linked_list.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        self.prev = prev
+
+class doubly_linked_list(object):
+    # Use items = doubly_linked_list()
+    def __init__(self):
+        self.head = None
+        self.tail = None
+        self.count = 0
+
+    def append_item(self, data):
```
Comment (codesankalp) [inline dsalgo/doubly_linked_list.py:15]: use append only.

File: dsalgo/doubly_linked_list.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            self.head = new_item
+            self.tail = self.head
+        else:
+            new_item.prev = self.tail
+            self.tail.next = new_item
+            self.tail = new_item
+
+        self.count += 1
+
+    def insert_start(self, data):
```
Comment (codesankalp) [inline dsalgo/doubly_linked_list.py:28]: use prepend here.

File: dsalgo/doubly_linked_list.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        
+        self.count += 1
+
+
+    def print_foward(self):
+        #  to print the output use items.print_foward()
+        for node in self.iter():
+            print(node)
+            
+    def iter(self):
```
Comment (codesankalp) [inline dsalgo/doubly_linked_list.py:44]: make it a private method.

**Review by codesankalp — CHANGES_REQUESTED** [review]
Good work @dennis-pg .
Please do the mentioned changes and this or will be ready to merge. :relaxed:

**Comment by codesankalp** [comment]
@dennis-pg are you still working on this issue?

**Comment by dennis-pg** [comment]
Sorry, got a bit caught up. Will finish up on this today

**Comment by dennis-pg** [comment]
Could you mention the issue numbers. Don't see any unresolved issues pertaining to doubly linked lists

**Comment by dennis-pg** [comment]
Have made the required changes

**Comment by codesankalp** [comment]
@dennis-pg I see there are many commits please make them into one (squash in one commit).

**Review by codesankalp — APPROVED** [review]
will add hacktober fest accepted label after you squash commits into one.

File: dsalgo/doubly_linked_list.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            yield item_val
+
+
+    def to_list(self):
+        current_node = self.head
+        arr = []
+        while current_node is not None:
+            arr.append(current_node.data)
+            current_node = current_node.next
+        return arr
```
Comment (devkapilbansal) [inline dsalgo/doubly_linked_list.py:63]: Add a new line here

File: tests/doubly_linked_list_tests.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        self.dll = dll()
+        for i in self.test_array:
+            self.dll.prepend(i)
+        self.assertEqual(self.dll.to_list(), self.test_array[::-1])
+
+    def test_size(self):
+        self.test_append()
+        self.assertEqual(self.dll.size(), len(self.test_array))
+
+if __name__ == '__main__':
```
Comment (devkapilbansal) [inline tests/doubly_linked_list_tests.py:28]: Remove this driver code

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@dennis-pg test file should start with the name 'test_' as prefix. Also, I saw that in some functions there is a single line while in some there are two. Please maintain consistency. Also, squash/rebase your commits into one as asked by @codesankalp

**Comment by dennis-pg** [comment]
Done

**Comment by devkapilbansal** [comment]
@dennis-pg I will review and test your PR soon. Till then please follow the commit guidelines mentioned in the contribution guide.

**Comment by codesankalp** [comment]
image
Also changed the styling because at someplace it is in the wrong format.
__Tested OK.__

Merging.

**Pull request merged**